CONTRIBUTING.md

```diff
@@ -0,0 +1,506 @@
+    idiomatically as possible.
+  * **Tested:** You must write (and pass) convincing [tests](#testing) for all
+    new or changed functionality.
+  * **Focused:** Your code should do what it's supposed to and nothing more.
+
+### Resolving an Open Issue
+[Resolving an Open Issue]: #resolving-an-open-issue
+
+If you spot an open issue that you'd like to resolve:
+
+
+  1. **First identify if there's a proposed solution to the problem.**
+
+     If there is, proceed to step 2. If there isn't, your first course of
+     action, before writing any code, is to propose a solution. To do so, leave
+     a comment describing your solution in the relevant issue. It's especially
+     useful to see test cases and hypothetical examples. This step is critical:
+     it allows us to identify and resolve concerns with a proposed solution
+     before anyone spends time writing code. It may also allow us to point you
+     in more efficient implementation directions.
+
+  2. **Write a failing test case you expect to pass after resolving the issue.**
+
+     If you can write proper tests cases that fail, do so (see [Testing]). If
+     you cannot, for instance because you're introducing new APIs which can't be
+     used until they exist, write a test case that mocks usage of those APIs. In
+     either case, allow the tests and mock examples to guide your progress.
+
+  3. **Write basic functionality, pass tests, and submit a PR.**
+
+     Think about edge cases to the problem and ensure you have tests for those
+     edge cases. Once your implementation is functionally complete, submit a PR.
+     Don't spend time writing or changing a bunch of documentation just yet.
+
+  4. **Wait for a review, iterate, and polish.**
+
+     If a review doesn't come in a few days, feel free to ping a maintainer.
+     Someone will review your PR eventually. Once they do, integrate their
+     feedback. If the PR solves the issue (which it should because you have
+     passing tests) and fits the project (which it should since you sought
+     feedback _before_ submitting), it will be _conditionally_ approved pending
+     final polish: documentation (rustdocs, guide docs), style improvements, and
+     testing. Your PR will then be merged.
+
+### Implementing an Unproposed Feature
+[Implementing an Unproposed Feature]: #implementing-an-unproposed-feature
+
+First and foremost, **please do not submit a PR that implements a new feature
+without first proposing a design and seeking feedback.** We take the addition of
+new features _very_ seriously because they directly impact usability.
+
+To propose a new feature, create a [new issue] and select "feature request."
+Follow the template, describing your solution in the _ideal solution_ or
+_additional context_ sections as appropriate. Note that certain classes of
+features require strong justification and exceedingly compelling reasons in
+order to be taken into consideration. These include features that:
```

Comment:
I think we should reword this, and provide more detail on what we want from a feature request (most of this is about what not suggest).

Some things I would look for in a feature request:
- Clear issue it's solving (e.g., A would now be possible, or A was very hard, but is now much easier)
- Code examples for what someone using the new feature would look like (and the current work around if relevant)
- Cohesive design
- Limits potential for confusion and foot guns
- As simple as possible (from the perspective of someone using Rocket)

Reply:
The feature request issue template already has fields for most of these. Do you think it doesn't suffice?

Reply:
It's been a little while since I've opened an issue, so I had to go re-read it. Although I think it covers what we want, I do think there is some value of repeating it here.

Alternatively, we could shorten this section a bit, and simply refer to the issue creation for both what we are, and are not looking for.